### PR TITLE
feature(core) Add GuzzleDownloader and dedicated exceptions to DownloaderInterface

### PR DESCRIPTION
**Pull request by rprzedzik**
# Description

Add Dedicated exception to DownloadeInterface
Add new Guzzle base GuzzleDownloader claas bas on DownloaderInterface

Issue #1749
## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

- [ ] Unit Tests
- [ ] Behat Test

# Checklist:
- [x] I have read the contribution requirements and fulfill them.
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: module/core/src/Infrastructure/Service/CurlDownloader.php [hunk 1]
```diff
@@ -9,16 +9,30 @@
 namespace Ergonode\Core\Infrastructure\Service;
 
 use Symfony\Component\HttpFoundation\Response;
+use Ergonode\Core\Infrastructure\Exception\DownloaderException;
+use Psr\Log\LoggerInterface;
+use Ergonode\Core\Infrastructure\Exception\FileNotFoundDownloaderException;
+use Ergonode\Core\Infrastructure\Exception\AccessDeniedDownloaderException;
+use Ergonode\Core\Infrastructure\Exception\BadRequestDownloaderException;
 
 class CurlDownloader implements DownloaderInterface
```
Comment (piotrkreft) [inline module/core/src/Infrastructure/Service/CurlDownloader.php:18]: I think this should become deprecated

File: module/core/src/Infrastructure/Service/GuzzleDownloader.php [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    /**
+     * @param Header[] $headers
+     *
+     * @throws DownloaderException
+     */
+    public function download(string $url, array $headers = []): string
+    {
+        $client = new Client();
```
Comment (piotrkreft) [inline module/core/src/Infrastructure/Service/GuzzleDownloader.php:36]: Should be injected from DI

File: module/core/src/Infrastructure/Service/GuzzleDownloader.php [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        try {
+            $response = $client->request('GET', $url, ['headers' => $this->mapHeaders($headers)]);
```
Comment (piotrkreft) [inline module/core/src/Infrastructure/Service/GuzzleDownloader.php:38]:
```suggestion
            $response = $client->get($url, ['headers' => $this->mapHeaders($headers)]);
```

**Review by piotrkreft — CHANGES_REQUESTED** [review]

File: UPGRADE-1.3.md [hunk 4]
```diff
@@ -11,3 +11,6 @@ Attribute:
 * The signature of constructor `DbalProductValueChangedEventProjector` has changed
 * The signature of constructor `DbalProductValueRemovedEventProjector` has changed
 * The signature of constructor `ProductWorkflowQuery` has changed
+
+Core
+* Remove nullable from `DownloaderInterface` return type
```
Comment (piotrkreft) [inline UPGRADE-1.3.md:16]:
```suggestion
* `DownloaderInterface::download` throws exception rather than returning `null` value
```

**Review by piotrkreft — CHANGES_REQUESTED** [review]

**Review by wfajczyk — DISMISSED** [review]

**Review by piotrkreft — APPROVED** [review]

**Review by wfajczyk — APPROVED** [review]

**Pull request merged**